Snowflake Connector issues fixed and added new features

naisonslalom opened the pull request:
---
### When submitting a connector, please make sure that you follow the requirements below, otherwise your PR might be rejected. We want to make you have a well-built connector, a smooth certification experience, and your users are happy :) 

If this is your first time submitting to GitHub and you need some help, please sign up for this [session](https://forms.office.com/pages/responsepage.aspx?id=KtIy2vgLW0SOgZbwvQuRaXDXyCl9DkBHq4A2OG7uLpdUMTFJWFFGVUxBNUFZQjZWRUdaOE5BMFkwNS4u). 

- [ ] I attest that the connector doesn't exist on the Power Platform today. I've verified by checking the pull requests in GitHub and by searching for the connector on the platform or in the documentation. 
- [x] I attest that the connector works and I verified by deploying and testing all the operations. 
- [x] I attest that I have added detailed descriptions for all operations and parameters in the swagger file.
- [x] I attest that I have added response schemas to my actions, unless the response schema is dynamic. 
- [x] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [x] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a connector icon.

If you are an Independent Publisher, you must also attest to the following to ensure a smooth publishing process:
- [ ] I have named this PR after the pattern of "Connector Name (Independent Publisher)" ex: HubSpot Marketing (Independent Publisher)
- [ ] Within this PR markdown file, I have pasted screenshots that show: 3 unique operations (actions/triggers) working within a Flow. This can be in one flow or part of multiple flows. For each one of those flows, I have pasted in screenshots of the Flow succeeding. 
- [ ] Within this PR markdown file, I have pasted in a screenshot from the Test operations section within the Custom Connector UI.
- [ ] If the connector uses OAuth, I have provided detailed steps on how to create an app in the readme.md.

naisonslalom committed:
Fix for HTTP 500 error

Fix for HTTP 500 error

naisonslalom committed:
Code clean up and removing Covert Action

naisonslalom committed:
Merge branch 'microsoft:dev' into dev

naisonslalom committed:
Update apiProperties.json

Added User-Agent header and updated connection parameters

naisonslalom committed:
Merge branch 'microsoft:dev' into dev

naisonslalom committed:
Update readme.md

Updated Audience URL with space holder
Added Connection credentials information

naisonslalom committed:
Merge branch 'dev' of https://github.com/naisonslalom/PowerPlatformConnectors into dev

naisonslalom committed:
Update apiDefinition.swagger.json

Added Covert action and updated schema

naisonslalom committed:
Updates from Karthik

Updates from Karthik

naisonslalom committed:
Minor updates in Regex and datatype update in schema

Minor updates in Regex
Datatype update in schema

naisonslalom committed:
Update script.csx

Updating the account regex to make it generic subdomain pattern

naisonslalom committed:
Adding 500 status code for Convert function

Adding 500 status code for Convert function

naisonslalom committed:
Updated regex validation for instance url

Updated regex validation for instance url

naisonslalom committed:
Updated schema for GetResults

Updated schema for GetResults
Updated regex validation

naisonslalom committed:
Updating escape sequence in regex

Updating escape sequence in regex

naisonslalom committed:
Update readme.md

Updated readme.md

naisonslalom committed:
Merge branch 'microsoft:dev' into dev

naisonslalom committed:
Update readme.md

Update readme.md

naisonslalom commented:
@microsoft-github-policy-service agree company="snowflake"


From: microsoft-github-policy-service[bot] ***@***.***>
Sent: Friday, February 16, 2024 2:26 PM
To: microsoft/PowerPlatformConnectors ***@***.***>
Cc: Naison Garvasis ***@***.***>; Mention ***@***.***>
Subject: Re: [microsoft/PowerPlatformConnectors] Snowflake Connector issues fixed and added new features (PR #3247)

@naisonslalom please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information. @microsoft-github-policy-service agree [company="{your company}"] Options: (default - no company


@naisonslalom<https://urldefense.com/v3/__https:/github.com/naisonslalom__;!!Jrgwm_wKFTPpVg!IUedqP7BBaZj1J5aLQ70u-CuA9ruLIuLSMwV9So3ZjKblKjZSMCBCP7Ivo4F-M4_jj28s9BACNAkpSHCFpLNiOBSWLbo$> please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.

@microsoft-github-policy-service agree [company="{your company}"]

Options:

  *   (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.

@microsoft-github-policy-service agree

  *   (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.

@microsoft-github-policy-service agree company="Microsoft"
Contributor License Agreement
Contribution License Agreement

This Contribution License Agreement (“Agreement”) is agreed to by the party signing below (“You”),
and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your
contributions to Microsoft open source projects. This Agreement is effective as of the latest signature
date below.

  1.  Definitions.
“Code” means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to Microsoft under this Agreement.
“Project” means any of the projects owned or managed by Microsoft and offered under a license
approved by the Open Source Initiative (www.opensource.org<https://urldefense.com/v3/__http:/www.opensource.org__;!!Jrgwm_wKFTPpVg!IUedqP7BBaZj1J5aLQ70u-CuA9ruLIuLSMwV9So3ZjKblKjZSMCBCP7Ivo4F-M4_jj28s9BACNAkpSHCFpLNiNpsUCbd$>).
“Submit” is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”
“Submission” means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.
  2.  Your Submission. You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.
  3.  Originality of Work. You represent that each of Your Submissions is entirely Your original work.
Should You wish to Submit materials that are not Your original work, You may Submit them separately
to the Project if You (a) retain all copyright and license information that was in the materials as You
received them, (b) in the description accompanying Your Submission, include the phrase “Submission
containing materials of a third party:” followed by the names of the third party and any licenses or other
restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
guidelines concerning Submissions.
  4.  Your Employer. References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.
  5.  Licenses.

  *   Copyright License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
parties.
  *   Patent License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.
  *   Other Rights Reserved. Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

  1.  Representations and Warranties. You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
  2.  Notice to Microsoft. You agree to notify Microsoft in writing of any facts or circumstances of which
You later become aware that would make Your representations in this Agreement inaccurate in any
respect.
  3.  Information about Submissions. You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.
  4.  Governing Law/Jurisdiction. This Agreement is governed by the laws of the State of Washington, and
the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
defenses of lack of personal jurisdiction and forum non-conveniens.
  5.  Entire Agreement/Assignment. This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/microsoft/PowerPlatformConnectors/pull/3247*issuecomment-1949194365__;Iw!!Jrgwm_wKFTPpVg!IUedqP7BBaZj1J5aLQ70u-CuA9ruLIuLSMwV9So3ZjKblKjZSMCBCP7Ivo4F-M4_jj28s9BACNAkpSHCFpLNiPGrsU72$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/BFPBUTW3HGDZ7VGTJ5XXMELYT6XFRAVCNFSM6AAAAABDMRFDE2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNBZGE4TIMZWGU__;!!Jrgwm_wKFTPpVg!IUedqP7BBaZj1J5aLQ70u-CuA9ruLIuLSMwV9So3ZjKblKjZSMCBCP7Ivo4F-M4_jj28s9BACNAkpSHCFpLNiJho_3Sh$>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

raanandm commented on certified-connectors/snowflake/readme.md:
```diff
@@ -25,12 +25,19 @@ create security integration connector
        external_oauth_issuer = '<AZURE_AD_ISSUER>'     
-       external_oauth_jws_keys_url = 'https://login.windows.net/common/discovery/keys'
-       external_oauth_audience_list = ('https://analysis.usgovcloudapi.net/powerbi/connector/snowflake')
+       external_oauth_jws_keys_url = '<Keys URL from step C>'
```
Step 3?

vmanoharas replied:
Hello @raanandm

May I know the reason that why you reviewed this PR?

naisonslalom replied:
@vmanoharas  - We have been working with @raanandm to fix some security compliance issues in this connector based on a Feb16th deadline given by Microsoft. @raanandm was trying to expedite the process.

naisonslalom replied:
Issue has been fixed and committed.

naisonslalom committed:
Update readme.md

Fixed review comments about step C

vmanoharas commented:
Thanks @naisonslalom. 

Is this ready for the review now?

naisonslalom commented:
> Thanks @naisonslalom.
> 
> Is this ready for the review now?

@vmanoharas - Yes, it's ready for the review

vmanoharas commented on certified-connectors/snowflake/apiProperties.json:
```diff
@@ -5,7 +5,6 @@
           "identityProvider": "aad",
-          "clientId": "54fbf434-4502-41d8-96a6-8e4fbe598e98",
           "scopes": [],
           "redirectMode": "Global",
```
Please modify the redirectMode as "GlobalPerConnector, please refer the below link. And kindly remove the redirectUrl, team will guide you to update this url in your client app during certification. 

https://learn.microsoft.com/en-us/connectors/custom-connectors/certification-submission#step-5-submit-your-connector-for-deployment

naisonslalom replied:
Hi @vmanoharas, implemented review comments, please verify.

vmanoharas reviewed (commented):
Hello @naisonslalom, 

Could you please resolve review comments and resolve the below apiProperties.json validation error?

Error - ##[error]@{level=Error; message=JSON does not match any schemas from 'anyOf'.; valuePath=line: 6 position: 26 path: '[0].properties.connectionParameters.token.oAuthSettings'; value=}

Please refer the schema link here to resolve the above error - https://github.com/microsoft/PowerPlatformConnectors/blob/dev/schemas/paconn-apiProperties.schema.json

naisonslalom committed:
Implemented review comments

Updated redireceMode as GlobalPerConnector
Updated redirectUrl as empty string
Added clientId and clientSecret with default null value

vmanoharas commented on certified-connectors/snowflake/apiProperties.json:
```diff
@@ -5,9 +5,11 @@
-          "redirectMode": "Global",
-          "redirectUrl": "https://global.consent.azure-apim.net/redirect",
+          "redirectMode": "GlobalPerConnector",
+          "redirectUrl": "",
```
Hello @naisonslalom, 

Please remove redirectUrl completely, as this is not needed.

naisonslalom replied:
@vmanoharas, done, please verify

vmanoharas reviewed (commented):
Hi @naisonslalom,

Can you please resolve one more comment?

naisonslalom committed:
Update apiProperties.json

Removed redirecturl

vmanoharas reviewed (approved):
Hello @naisonslalom 
I hope you are doing well.
Congratulations, your pull request is approved and merged. Please [submit the files to ISV Studio](https://docs.microsoft.com/en-us/connectors/custom-connectors/submit-for-certification#submit-to-isv-studio) or [update your submission](https://learn.microsoft.com/en-us/connectors/custom-connectors/submit-for-certification#update-submission-prior-to-certification) using your latest commit ID, please allow us up to 1-2 weeks to review your submission, and our Engineers will notify you in the “Activity Control” area in ISV Studio. Please [make sure that the icon meets all the requirements](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-submission#design-an-icon-for-your-connector), that it’s size is 230x230, icon background is not white or transparent and it matches the iconBrandColor property in apiProperties.json file.
After making the submission in ISV Studio, please attach an intro.md file in the activity control section. This file is required and it's different from readme.md. You can [find an example of intro.md file on this page](https://docs.microsoft.com/en-us/connectors/custom-connectors/submit-for-certification#submit-to-isv-studio) at step 6.
Please [Create an environment in the Preview region](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-testing#create-an-environment-in-the-preview-region). You will use this environment later to [test your connector](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-testing#test-your-connector) after Microsoft is done with the functional verification for your connector.
We expect all tests to be completed within 1-2 weeks.
If your connector passes all tests, the deployment process begins, and it typically takes up to 3 to 4 weeks to deploy your connector to all of our regions.
Please let me know if you have any questions or run into any problems while creating or updating your submission.
Thank you very much for working with us.
